File: Binary Search/1608. Special Array With X Elements Greater Than or Equal X/1608.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        n=len(nums)
        l,r=0,n

        def bisectleft(target):
            l,r=0,n

            while l < r:
                mid = (l + r) // 2
                if nums[mid] >= target:
                    r = mid
                elif nums[mid] < target:
                    l = mid + 1
            return l
        
        while l <= r:
            mid = (l + r) // 2
            pos = bisect_left(nums, mid)
            # pos = bisectleft(mid)
            count = n - pos

            if count == mid:
                return mid
            elif count > mid:
                l = mid + 1
            elif count < mid:
                r = mid - 1
        
        return -1

File: Binary Search/1608. Special Array With X Elements Greater Than or Equal X/1608.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","fast-math","march=native","lto","omit-frame-pointer")
class Solution {
public:
    int specialArray(vector<int>& nums) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        int n=nums.size();
        sort(nums.begin(),nums.end());

        int l=0;
        int r=n;

        while (l<=r) {
            int mid=(l+r)/2;
            auto itr=lower_bound(nums.begin(),nums.end(),mid);
            int count=n-distance(nums.begin(),itr);
            // int pos=binSearch(nums,n, mid);
            // int count=n-pos;
            
            if (count==mid) {
                return mid;
            }
            else if (count>mid) {
                l=mid+1;
            }
            else if (count<mid) {
                r=mid-1;
            }
        }
        return -1;
    }

private:
    const int binSearch(const vector<int>& nums, const int& n, const int& target) const {
        int l=0;
        int r=n;
        while (l<r) {
            int mid=(l+r)/2;
            if (nums[mid]>=target) r=mid;
            else if (nums[mid]<target) l=mid+1;
        }
        return l;
    }
};